core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala

```diff
@@ -52,43 +51,84 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
+          case None =>
+            info(s"Transaction topic partition for $transactionalId may likely has emigrated, as the corresponding metadata do not exist in the cache" +
+              s"any more; cancel sending transaction markers $txnMarker to the brokers")
+
+            // txn topic partition has likely emigrated, just cancel it from the purgatory
+            txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
+
+          case Some(epochAndMetadata) =>
+            val txnMetadata = epochAndMetadata.transactionMetadata
+
+            if (epochAndMetadata.coordinatorEpoch != txnMarker.coordinatorEpoch) {
+              // coordinator epoch has changed, just cancel it from the purgatory
+              info(s"Transaction coordinator epoch for $transactionalId has changed from ${txnMarker.coordinatorEpoch} to " +
+                s"${epochAndMetadata.coordinatorEpoch}; cancel sending transaction markers $txnMarker to the brokers")
+
+              txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
+            }
+
+            txnMetadata synchronized {
+              for ((topicPartition: TopicPartition, error: Errors) <- errors) {
+                error match {
+                  case Errors.NONE =>
+
+                    txnMetadata.removePartition(topicPartition)
+
+                  case Errors.CORRUPT_MESSAGE |
+                       Errors.MESSAGE_TOO_LARGE |
+                       Errors.RECORD_LIST_TOO_LARGE |
+                       Errors.INVALID_REQUIRED_ACKS => // these are all unexpected and fatal errors
+
+                    throw new IllegalStateException(s"Received fatal error ${error.exceptionName} while sending txn marker for $transactionalId")
+
+                  case Errors.UNKNOWN_TOPIC_OR_PARTITION |
+                       Errors.NOT_LEADER_FOR_PARTITION |
+                       Errors.NOT_ENOUGH_REPLICAS |
+                       Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND => // these are retriable errors
+
+                    info(s"Sending $transactionalId's transaction marker for partition $topicPartition has failed with error ${error.exceptionName}, retrying " +
+                      s"with current coordinator epoch ${epochAndMetadata.coordinatorEpoch}")
+
+                    retryPartitions += topicPartition
+
+                  case Errors.INVALID_PRODUCER_EPOCH |
+                       Errors.TRANSACTION_COORDINATOR_FENCED => // producer or coordinator epoch has changed, this txn can now be ignored
+
+                    info(s"Sending $transactionalId's transaction marker for partition $topicPartition has permanently failed with error ${error.exceptionName} " +
+                      s"with the current coordinator epoch ${epochAndMetadata.coordinatorEpoch}; cancel sending any more transaction markers $txnMarker to the brokers")
+
+                    txnMarkerChannelManager.removeMarkersForTxnId(transactionalId)
```

Comment:
`TransactionMetadata` should be removed by the emmigration handler thread, and in the future the PID expiration scheduler's thread. Other handling logic should never remove the entry, but just to update the entry in-place.

And about returning here: good catch! yeah we should not call `completeSendMarkersForTxnId` in this case. Will fix now.